platform/engineering/design-docs/2021-01-13-web-component-library.md

```diff
@@ -0,0 +1,185 @@
+
+#### Loading Web Components on va.gov
+
+Once in `vets-website`, the setup code for the WC library will be in its own entry file that is loaded on each page. This code will be responsible for applying polyfills for older browsers as well as defining the Web Components.
+
+In addition, this script will be one of the first loaded in order to ensure that the Components will be defined in the browser before they are rendered on the page. This will be done by adding a `<script>` entry towards the top of [the `<head>` tag that gets loaded on every page](https://github.com/department-of-veterans-affairs/vets-website/blob/dc5018065eccf073d172097a19922755126c98a3/src/site/includes/header.html#L16)
+
+For further optimization, the Components will be lazy-loaded. If our Web Component library ends up having hundreds of Components, this could cause a significant performance impact if we defined these every time a new user landed on a va.gov page. Instead, the Components will only _actually_ be defined if they appear on the page. [Stencil has a useful article that talks about their lazy loading approach](https://stenciljs.com/blog/how-lazy-loading-web-components-work).
+
+### Code Location
+
+The code will live in the `component-library` repo on a dedicated orphan branch named `web-components`. This code will be used in `vets-website` via a dependency in `package.json`
+
+### Testing Plan
+
+Each Component will have:
+
+- Unit and E2E tests written using Jest
+- One or more stories in Storybook to validate appearance/behavior
+
+As of mid-January 2021, Stencil provides an experimental screenshot visual diff testing functionality. If this becomes stable, we will consider adopting it.
+
+### Logging
+
+N/A
+
+### Debugging
+
+Storybook will be a useful tool for debugging and development. Stencil also comes with its own development server which can be used.
+
+There will be no special provisions for debugging provided for consumers of the WC library.
+
+### Caveats
+
+Even though Web Components are already a feature provided by the browser, this is a feature that the VA Design System hasn't yet taken advantage of. By introducing this "new" technology, there is a possibility of this just turning into technical debt if it exists alongside the current `component-library`, `formation` component styles, and Liquid templates for components. In order to avoid this, we want to be agressive about this migration and get out of the transitionary period as quickly and feasibly as we can.
```

Comment:
From the template, this section is for
> Gotchas, differences between the design and implementation, other potential stumbling blocks for users or maintainers, and their implications and workarounds. Unless something is known to be tricky ahead of time, this section will probably start out empty.

What you've got here seems more like a risk to me. What do you think?

Reply:
Or right! I meant to leave a comment on this since I had a question about this exact thing :grin: . I put it in Caveats because:

> other potential stumbling blocks for users or maintainers, and their implications

I'll go ahead and move it to Risks though.